udata_hydra/analysis.py

```diff
@@ -75,49 +74,76 @@ async def process_resource(check_id: int) -> None:
+        tmp_file = None
+        try:
+            tmp_file = await download_resource(url, headers)
+        except IOError:
+            dl_analysis["analysis:error"] = "File too large to download"
+        else:
+            # Get file size
+            dl_analysis["analysis:filesize"] = os.path.getsize(tmp_file.name)
+            # Get checksum
+            dl_analysis["analysis:checksum"] = compute_checksum_from_file(tmp_file.name)
+            # Check if checksum has been modified if we don't have other hints
+            change_analysis = (
+                await detect_resource_change_from_checksum(resource_id, dl_analysis["analysis:checksum"])
+                or {}
+            )
+            # TODO: this never seems to output text/csv, maybe override it later
+            dl_analysis["analysis:mime-type"] = magic.from_file(tmp_file.name, mime=True)
+        finally:
+            if tmp_file:
+                os.remove(tmp_file.name)
+            await update_check(check_id, {
+                "checksum": dl_analysis.get("analysis:checksum"),
+                "analysis_error": dl_analysis.get("analysis:error"),
+                "filesize": dl_analysis.get("analysis:filesize"),
+                "mime_type": dl_analysis.get("analysis:mime-type"),
+            })
+
+    has_changed_over_time = await detect_has_changed_over_time(change_analysis, resource_id, check_id)
+
+    change_analysis.update(dl_analysis)
+    if has_changed_over_time or (is_first_check and change_analysis):
         await send(
             dataset_id=dataset_id,
             resource_id=resource_id,
-            document={
-                "analysis:error": error,
-                "analysis:filesize": filesize,
-                "analysis:mime-type": mime_type,
-                "analysis:checksum": sha1,
-                **change_analysis,
-            }
+            document=change_analysis
         )
-        await update_check(check_id, {
-            "checksum": sha1,
-            "analysis_error": error,
-            "filesize": filesize,
-            "mime_type": mime_type
-        })
 
 
-async def detect_resource_change_from_checksum(resource_id, new_checksum):
+async def detect_has_changed_over_time(change_analysis, resource_id, check_id) -> bool:
+    """
+    Determine if our detected last modified date has changed since last check
+    because some methods (eg last-modified header) do not embed this
+    """
+    has_changed_over_time = False
+    last_modified = change_analysis.get("analysis:last-modified-at")
+    if last_modified:
+        q = """
+        SELECT detected_last_modified_at
+        FROM checks
+        WHERE resource_id = $1 AND detected_last_modified_at IS NOT NULL
+        ORDER BY created_at DESC
+        LIMIT 1
+        """
+        pool = await context.pool()
+        async with pool.acquire() as conn:
+            res = await conn.fetchrow(q, resource_id)
+            if res and res["detected_last_modified_at"] != last_modified:
```

Comment:
Shouldn't we cast last_modified to isoformat before comparing? (Not sure why it doesn't show in tests though)

Reply:
Nope, the ORM returns datetime objects for timestamp columns :-)

Reply:
My bad, I mean we're comparing datetime objects and isoformat string here?

Reply:
Hum yes 😱 And it's working 🤷

Reply:
Fixed!